### ISSUE TITLE
CORS Error

### ISSUE THREAD

**Issue by edolyne** [markhuot/craftql#51]
Not quite sure how you want to handle this, but when calling the `/api` endpoint from another domain a CORS error is received and the request is blocked. Right now, this is just when I'm working locally on a project, I haven't tried it with a production setup.

```
Failed to load https://mycraftdomain.test/api: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:3000' is therefore not allowed access. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

I did find a potential solution to the issue here: https://stackoverflow.com/questions/45904060/request-graphql-api-running-on-different-port-from-next-js-app-in-development-mo

**Comment by timkelty** [comment]
I think it's because your `siteUrl` setting isn't matching your request url (localhost:3000). Try making them match.

@markhuot You _could_ consider checking for `devMode` and do this to alleviate this kind of trouble in dev:
`Craft::$app->getResponse()->getHeaders()->set('Access-Control-Allow-Origin', '*');`

Not sure if thats a great idea though.

**Comment by markhuot** [comment]
@edolyne, you can set "Allowed Origins" in a `config/craftql.php` file.

https://github.com/markhuot/craftql/blob/945e17504acf1eaf824b2e4c452e6717def96017/src/Controllers/ApiController.php#L59-L69

To do so, create `config/craftql.php` and add,

```php
<?php

return [
    'allowedOrigins' => [
        'localhost', // or '*', as Tim mentions
    ]
];
```

**Comment by edolyne** [comment]
@timkelty My goal would be to have the ability for Craft to live at subdomain, with requests being accepted from an outside server.

**Comment by edolyne** [comment]
@markhuot That's perfect!

**Comment by markhuot** [comment]
I'm going to close this. If you're still having trouble or need some help getting the `allowedOrigins` set up feel free to re-open it and let me know.

**Comment by ads1018** [comment]
@markhuot I'm still getting a `No 'Access-Control-Allow-Origin'` error even after I added `craftql.php` to my config. Any idea why?

**Comment by stursby** [comment]
Hey @markhuot, I know this is closed, but didn't feel the need to open a new issue.

So I've got Craft 3 (3.0.0-RC17.1) running locally and then have a [Nuxt.js](https://nuxtjs.org/) site running on localhost:3000, and using the [@nuxtjs/apollo](https://github.com/nuxt-community/apollo-module) module.

I'll preface this by saying I think this is likely a SSR/CORS issue, but I've added your recomendation in `/config/craftql.config`

```php
<?php

return [
    'allowedOrigins' => [
        'localhost'
    ]
];

```

Here's a sample GraphQL query for the homepage

```gql
query {
	entry(section:[homepage]) {
    ... on Homepage {
      id
      title
    }
  }
}

```

And then in my Nuxt page

```vue
<template>
  <div>
    <h1>Homepage</h1>
    <nuxt-link to="/about">About</nuxt-link>
    <pre>{{ entry }}</pre>
  </div>
</template>

<script>
import homepage from '~/apollo/queries/homepage'

export default {
  name: 'Homepage',

  apollo: {
    entry: {
      prefetch: true,
      query: homepage
    }
  }
}
</script>
```

This works fine on first load (from the server, and also because `prefetch: true`, I think?)

![image](https://user-images.githubusercontent.com/527849/38318868-0bb94416-37f6-11e8-9e97-6e23b15a6f32.png)

But the second I try and navigate to /about (and load the `about.gql` query)... I get the following console error

![image](https://user-images.githubusercontent.com/527849/38318930-31daa57c-37f6-11e8-858b-a67bbe8dd04d.png)

Again, hopefully I'm missing something simple, but I thought I had followed all the right steps. Any thoughts would be much appreciated, and thank you for the amazing library!! 🙌🏻

**Comment by markhuot** [comment]
Try, 

```php
return [
    'allowedOrigins' => [
        'localhost:3000'
    ]
];
```

I _think_ the port is a part of the host name, typically.

**Comment by stursby** [comment]
Adding the port resulted in the same CORS error... however, I just tried using `*` instead, and I no longer get the CORS error, but the first time I a new page the data _is not_ loaded, but if I navigate back to `/` then back to `/about` ... it works 🤔

And I have a new console error

![image](https://user-images.githubusercontent.com/527849/38319282-141789be-37f7-11e8-80c0-a96f25990fe1.png)

Wondering if I need to check which context (server vs client) Nuxt is running in and `prefetch` according?

**Comment by markhuot** [comment]
Thanks for the heads up @stursby. I'll look at `localhost` v.s. `*` too.

**Comment by exophunk** [comment]
Hi markhuot,

I've run into the same issue. After I debugged the ApiController.php, I think the issue is because of this:

https://github.com/markhuot/craftql/blob/945e17504acf1eaf824b2e4c452e6717def96017/src/Controllers/ApiController.php#L64

The check `in_array($origin, $allowedOrigins)` compare the `$origin` & `$allowedOrigin` 1 to 1. So the `$origin` is `http://localhost:3000` in my case.

When adding 
```
return [
    'allowedOrigins' => [
        'http://localhost:3000'
    ]
];
```

to `configs/craftql.php`, it works fine. 

Altough I think it should strip out the protocol `http://`, that would be better.

**Comment by markhuot** [comment]
So, my thinking behind including `http://` on there is that the actual header sent with in the response would include `http://` too, e.g., `Access-Control-Allow-Origin: https://foo.bar`.

Also, including `https://` would allow you to only allow cross-site requests for secure connections…?

**Comment by bryanjhickey** [comment]
I am have the same issue:

> Access to XMLHttpRequest at 'https://api.galleries.test/api' from origin 'https://galleries.test' has been blocked by CORS policy: Request header field x-requested-with is not allowed by Access-Control-Allow-Headers in preflight response.

I get a successful api response with `curl` passing the `--insecure` flag but get a CORS warning when attempting to access from the site

I have also tried running the site at `http://localhost:8080` 

I have set my `craftql.php` to
```
<?php

return [
    'allowedOrigins' => [
        '*', // or '*', as Tim mentions
    ]
];
```

I also tried specifying the site URL (and port for localhost) but same results

I wondered if it might be an issue with the security cert and Laravel Valet so switched to 'http://' but still received the same CORS warning.
